File: machines/calcifer/kal/src/time.rs
use chrono::Timelike;
use thiserror::Error;

#[derive(
    Debug,
    Copy,
    Clone,
    Eq,
    PartialEq,
    Ord,
    PartialOrd,
    derive_more::Add,
    derive_more::Sub,
    derive_more::Mul,
)]
pub struct Time(u32);

impl std::ops::Div for Time {
    type Output = f64;
    fn div(self, rhs: Time) -> Self::Output {
        (self.0 as f64) / (rhs.0 as f64)
    }
}

#[derive(Error, Debug)]
pub enum TimeError {
    #[error("Wrong time")]
    Wrong,
}

pub type TimeResult = Result<Time, TimeError>;

impl Time {
    pub const MIN: Self = Self(0);
    pub const MAX: Self = Self(24 * 60);

    pub fn now() -> Self {
        chrono::Local::now().into()
    }
    pub fn from_minutes(minutes: u32) -> TimeResult {
        if minutes <= 24 * 60 {
            Ok(Self(minutes))
        } else {
            Err(TimeError::Wrong)
        }
    }
    pub fn from_hours(hours: f64) -> TimeResult {
        if 0.0 <= hours && hours <= 24.0 {
            Ok(Self::from_hours_unchecked(hours))
        } else {
            Err(TimeError::Wrong)
        }
    }
    pub fn from_hours_unchecked(hours: f64) -> Time {
        Self((hours * 60.0) as u32)
    }
}

impl<T: Timelike> From<T> for Time {
    fn from(time: T) -> Self {
        Self(time.minute() + time.hour() * 60)
    }
}

impl std::fmt::Display for Time {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:02}H{:02}", self.0 / 60, self.0 % 60)
    }
}

File: machines/calcifer/kal/src/kal/time.py
from datetime import datetime
from zoneinfo import ZoneInfo


class Time:
    MIN: int = 0
    MAX: int = 24 * 60

    def __init__(self, value: int = 0):
        assert self.MIN <= value <= self.MAX
        self._value: int = value

    def __str__(self):
        return f"{self._value / 60:02}H{self._value % 60:02}"

    def json(self) -> str:
        return str(self._value)

    def __hash__(self) -> int:
        return self._value

    def __eq__(self, other) -> bool:
        if isinstance(other, int):
            other = Time(other)
        elif isinstance(other, float):
            other = Time.from_hours(other)
        elif not isinstance(other, Time):
            return False
        return self._value == other._value

    def __ge__(self, other) -> bool:
        if isinstance(other, int):
            other = Time(other)
        elif not isinstance(other, Time):
            return False
        return self._value >= other._value

    def __gt__(self, other) -> bool:
        if isinstance(other, int):
            other = Time(other)
        elif not isinstance(other, Time):
            return False
        return self._value > other._value

    def __le__(self, other) -> bool:
        if isinstance(other, int):
            other = Time(other)
        elif not isinstance(other, Time):
            return False
        return self._value <= other._value

    def __lt__(self, other) -> bool:
        if isinstance(other, int):
            other = Time(other)
        elif not isinstance(other, Time):
            return False
        return self._value < other._value

    def __sub__(self, other) -> int:
        if isinstance(other, int):
            other = Time(other)
        return self._value - other._value

    @classmethod
    def now(cls):
        tz = ZoneInfo("Europe/Paris")
        return cls.from_datetime(datetime.now(tz=tz))

    @classmethod
    def from_datetime(cls, dt: datetime):
        return cls(dt.hour * 60 + dt.minute)

    @classmethod
    def from_hours(cls, hours: float):
        return cls(int(hours * 60.0))
